text/0000-terminology-and-boundaries.md

```diff
@@ -0,0 +1,224 @@
+  particular protocol is not part of the firmament, as it is not relevant to a
+  hardware driver.
+- Firefox, which uses the desktop environment's IPC interface to create a window
+  onscreen, statically links to the Vulkan library for rendering, and uses the
+  Sockets firmament specification.
+
+**Example 3:** a cluster of [TALOS II 2U servers](https://www.raptorcs.com/content/TL2SV1/intro.html)
+being used as a machine learning laboratory might use:
+
+- A nucleus compiled for POWER9, that integrates with OpenBMC.
+- The official Nebulet kernel.
+- NVIDIA graphics drivers that implement the GPGPU firmament specification.
+- Ethernet drivers that implement the Data Plane firmament specification.
+- An implementation of the Sockets firmament.
+- TensorFlow for clusters ported to a Nebulet application that uses the GPGPU
+  and Socket firmaments.
+- Jupyter and the entire SciPy stack (lol) ported to Nebulet, using the Sockets
+  firmament and many ad-hoc application-layer IPC interfaces.
+
+Note that in all examples, we can re-use the Nebulet kernel and firmament
+interfaces.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The nucleus must provide a thin set of abstractions over the hardware, in
+particular the CPU. These abstractions must include:
+- Access and control over CPU power states, if applicable.
+- Access and control over CPU interrupts, if applicable.
+- Access and control of the MMU, if applicable.
+- Access and control over CPU registers if necessary for common drivers on
+  the target hardware.
+
+The nucleus can also implement part of a verified boot chain if such a feature
+is desirable. In this case, the nucleus will be verified by the bootloader and
+it will verify the Nebulet kernel and other startup data, such as initial
+ramdisk or commandline.
+
+The Nebulet kernel will use these functions to implement the core functionality
+of threading, scheduling, process isolation, and inter-process communication.
+
+A firmament specification will describe the IPC interfaces and datatypes that
+an application using that piece of hardware will interact with. The Data Plane
+firmament specification used in the above examples might be defined as follows
+(using Rust-like syntax):
+
+```rust
+// The supporting structs and enums are assumed to have
+// functions that can be called easily from the receiver side.
+// We'll need to flesh out the IPC system for this to be in any
+// way realistic.
+struct MacAddr([u8; 6]);
+enum Ethertype {
+    Arp,
+    Ipv4,
+    Ipv6,
+}
+
+struct EthernetPacket {
+    data: Bytes,
+}
+
+impl EthernetPacket {
+    pub fn dst(&self) -> &MacAddr;
+    pub fn src(&self) -> &MacAddr;
+    pub fn ethertype(&self) -> &Ethertype;
+    pub fn next_layer(&self) -> &L3Packet;
+}
+
+interface EthernetNic {
+    fn send(&mut self, pkt: EthernetPacket) -> Result<()>;
+    fn send_inplace(&mut self, func: impl Fn(&mut EthernetPacket) -> Result<()>) -> Result<()>;
```

Comment:
I'm curious how `send_inplace` would work. The user supplies it a function that takes a EthernetPacket reference?

Reply:
See link for an example of a similar function. The idea here is that the network driver can allocate a packet located inside its Tx queue, then the application modifies the packet.
https://github.com/libpnet/libpnet/blob/master/pnet_datalink/src/linux.rs#L231

Reply:
Oh interesting! I hadn't thought of that. It'd be difficult to get the packet into the other wasm linear memory without copying, but could be possible through page flipping.